pkg/stream/connpool/msgconnpool/connpool.go

```diff
@@ -0,0 +1,347 @@
+		} else {
+			log.DefaultLogger.Warnf("[connpool] reconnect failed %v times, ac : %p",
+				ac.connectTryTimes, ac)
+		}
+
+		return
+	}
+
+	// atomic store, avoid partial write
+	atomic.StorePointer((*unsafe.Pointer)(unsafe.Pointer(&ac.host)), unsafe.Pointer(&createConnData))
+
+	log.DefaultLogger.Infof("[connpool] reconnect succeed after %v tries, host %v, ac %p, host: %v",
+		ac.connectTryTimes, ac.host.Host.AddressString(), ac, ac.host)
+
+	// if pool was destroyed, but connection was connected
+	// we need to close it
+	if atomic.LoadUint64(&ac.pool.destroyed) == 1 {
+		ac.host.Connection.Close(api.NoFlush, api.LocalClose)
+		return
+	}
+
+	// read filters
+	for _, rf := range ac.pool.readFilters {
+		ac.host.Connection.FilterManager().AddReadFilter(rf)
+	}
+
+	// set the new heartbeat
+	ac.setHeartBeater(ac.pool.heartBeatCreator())
+
+	// clear retry times
+	ac.connectTryTimes = 0
+}
+
+// reconnect triggers connection to reconnect
+func (ac *activeClient) reconnect() {
+	if atomic.LoadUint64(&ac.pool.destroyed) == 1 {
+		return
+	}
+
+	if ac.connectTryTimes > ac.pool.connTryTimes {
+		log.DefaultLogger.Warnf("[connpool] retry time exceed pool config %", ac.pool.connTryTimes)
+		return
+	}
+
+	var idx = ac.connectTryTimes
+	if idx >= len(ac.reconnectBackoff) {
+		idx = len(ac.reconnectBackoff) - 1
+	}
+
+	utils.GoWithRecover(func() {
+		time.Sleep(ac.reconnectBackoff[idx])
+		ac.reconnectLock.Lock()
+		defer ac.reconnectLock.Unlock()
+
+		if _, ok := ac.pool.isActive(); ok {
+			return
+		}
+
+		// close previous conn
+		// if ac.host != nil {
+			//ac.host.Connection.Close(api.NoFlush, api.LocalClose)
+		// }
+
+		ac.connectTryTimes++
+
+		if atomic.LoadUint64(&ac.pool.destroyed) == 1 {
+			// if pool was exited, then stop
+			return
+		}
+
+		ac.initConnection()
+
+	}, func(r interface{}) {
+		log.DefaultLogger.Warnf("[connpool] reconnect failed, %v, host: %v", r, ac.host.Host.AddressString())
+	})
+
+}
+
+// removeFromPool removes this client from connection pool
+func (ac *activeClient) removeFromPool() {
+	p := ac.pool
+	p.clientMux.Lock()
+	defer p.clientMux.Unlock()
+
+	p.client = nil
+}
+
+// setHeartBeater set the heart beat for an active client
+func (ac *activeClient) setHeartBeater(hb KeepAlive) {
+	// clear the previous keepAlive
+	if ac.keepAlive != nil && ac.keepAlive.keepAlive != nil {
+		ac.keepAlive.keepAlive.Stop()
+		ac.keepAlive = nil
+	}
+
+	ac.keepAlive = &keepAliveListener{
+		keepAlive: hb,
+		conn:      ac.host.Connection,
+	}
+
+	// this should be equal to
+	// ac.codecClient.AddConnectionEventListener(ac.keepAlive)
+	ac.host.Connection.AddConnectionEventListener(ac.keepAlive)
+}
+
+// keepAliveListener is a types.ConnectionEventListener
+type keepAliveListener struct {
+	keepAlive KeepAlive
+	conn      api.Connection
+}
+
+// OnEvent impl types.ConnectionEventListener
+func (l *keepAliveListener) OnEvent(event api.ConnectionEvent) {
+	if event == api.OnReadTimeout && l.keepAlive != nil {
```

Comment:
you should Stop keepalive if event is close ？

Reply:
keepalive will be cleared by initConnection

Reply:
if initconnection failed, it will not be cleared.

Reply:
now the heartbeat can be cleared at the beginning in initConnection

Reply:
if  autoReconnectWhenClose is fasle, heartbeat can not be cleared.

Reply:
done